File: packages/bot/src/utils/generators.js
import { Collection } from 'discord.js';
import { regexImageAndDescription } from './regex.js';

/**
 * @example
 * 
 * const { nameGenerator } = require('./utils/generator');
 * const bookName = 'The_Book_Name';
 * 
 * console.log(nameGenerator(bookName));
 * // Output: The Book Name
 * 
 * @description Removes special characters, numbers and underscores from the book name
 * @param { string } bookName - The name of the book
 * @returns { string } - The clean and readable name of the book
 */

export const nameGenerator = (bookName) => {
    bookName = bookName.replace(/[-\d\W_]+/g, ' ');
    bookName = bookName.replace("pdf", '')
    bookName = bookName.trim();
    return bookName;
}

/**
 * @typedef { import("../@types/types.js").Content } Content
 * @description Parse the message to get the image and description
 * @param { Collection } message - The message to be parsed
 * @returns { Content } The image and description of the message
 */

export const imageAndDescriptionGenerator = (message) => {
    const { content } = message;
    const matches = content.match(regexImageAndDescription);

    if (matches) {
        const contentImage = matches[1].trim();
        const contentDescription = matches[2].trim();
        return { image: contentImage, description: contentDescription };
    }

    return null;
}

File: packages/got/src/utils/generators.go
package utils

import (
	"regexp"
	"strings"
)

type Content struct {
	Image       string
	Description string
}

// NameGenerator takes a string and returns a modified version of the string.
// It replaces all non-alphanumeric characters with spaces and removes the
// string "pdf" from the end of the string (if present). It also trims all
// whitespace from the string.
func NameGenerator(bookName string) string {
	re := regexp.MustCompile(`[-\d\W_]+`)

	bookName = re.ReplaceAllString(bookName, " ")
	bookName = strings.Replace(bookName, "pdf", "", -1)
	bookName = strings.TrimSpace(bookName)

	return bookName
}

// ImageAndDescriptionGenerator extracts image URL and description from a message string.
// The message should contain "image: <url>\ndescription: <message>" format.
// Returns a pointer to a Content struct with the extracted image and description, or nil if no match is found.
func ImageAndDescriptionGenerator(message string) *Content {
	regexImageAndDescription := regexp.MustCompile(`image: (.*?)\ndescription: (.*)`)

	matches := regexImageAndDescription.FindStringSubmatch(message)

	if matches != nil {
		contentImage := strings.TrimSpace(matches[1])
		contentDescription := strings.TrimSpace(matches[2])
		return &Content{Image: contentImage, Description: contentDescription}
	}

	return nil
}
